net: initial import of the ZEP protocol

authmillenon opened the pull request:
Main idea is to use this with native as "fully" supported IEEE 802.15.4 device.

This modules is both a `ng_netdev` and `ng_netapi` module. It speaks through UDP via `ng_netapi`, but imitates an IEEE 802.15.4 device through `ng_netdev` to be indistinguishable from normal devices.

~~Depends on #2430.~~ (merged)

Fixes #163

haukepetersen commented:
Did you choose the naming yourself or is this a (well-known) protocol of some sort? Havn't heard about it so far...

Also I don't seem to understand this implementation. What is the duality of netdev and netapi about? And why does the protocol need it's own nettype entry? If it is running on top of UDP, netdev and the nettype should not be needed!

OlegHahm commented:
It's a Wireshark specific header: https://www.wireshark.org/docs/dfref/z/zep.html It's on RIOT's wishlist for a long time: https://github.com/RIOT-OS/RIOT/issues/163

authmillenon commented:
> Also I don't seem to understand this implementation. What is the duality of netdev and netapi about? And why does the protocol need it's own nettype entry? If it is running on top of UDP, netdev and the nettype should not be needed!

Yes, nettype might be replaceable with port `demux_ctx`. The `netdev` inclusion is explained in the description: this module will mock a netdev-compatible IEEE 802.15.4 device. For the major use-case (debugging/testing 6LoWPAN) the stack would look like that:

```
          +-------------+
          |             |
+-------------------+   |
|        ZEP        |   |
+-------------------+   |
|        UDP        |   |
+-------------------+   |
|       IPv6        |   |
+---------+---------+   |
|  nomac  | 6LoWPAN |   |
+---------+---------+   |
| tapnet  |  nomac  |   |
+---------+---------+   |
     |    | zep_dev |---+
    TAP   +---------+
```

haukepetersen commented:
I am not completely sure if I understand it fully, but if it is the way I am thinking this is indeed useful. So +1 for this.

But there is still stuff missing in the implementation, right?

authmillenon commented:
Yes it is still WIP.

authmillenon commented:
I've done some work on this again, and it's based on UDP (#2430) now.

The basic behavior (send data frames and receive data frames) should work now. I reorder and document the stuff when I find some time, but in general I would add no more further functionally to the module for now.

authmillenon commented:
Rebased to current master

OlegHahm commented:
@Lotterleben, if you don't want to review just re-assign to me (or someone else). I'm definitely interested in reviewing, but my backlog is increasing by the minute...

Lotterleben commented:
@OlegHahm I'll have to look into ZEP first to do this one justice, so it might take me a while too... If somebody else is more familiar with the matter and has the time to review this PR, I'll gladly pass it on, but otherwise, maybe we can form a reviewing team together?

sgso commented:
Would I be eligible (in the team)? I know as much as you do about ZEP (close to nothing) but would be interested to have it for the sixlowpan-part of the coap plugtests...

authmillenon commented:
@sgso wouldn't that be more sensible to have this running on actual hardware? But sure to look at the output in wireshark it would be also helpful for you.

authmillenon commented:
Resources on ZEP are pretty rare by the way. Most of it I did via reverse engineering and by looking at the Wireshark dissector code of it, but here are the resources I used:

* [The official doc at the wireshark wiki](https://www.wireshark.org/docs/dfref/z/zep.html)
* [packet-zep.h](https://code.wireshark.org/review/gitweb?p=wireshark.git;a=blob;f=epan/dissectors/packet-zep.h;h=519b1a781a6e9a51e3f7bb9174edacedd5fff85e;hb=HEAD) from the wireshark repository
* [packet-zep.c](https://code.wireshark.org/review/gitweb?p=wireshark.git;a=blob;f=epan/dissectors/packet-zep.c;h=3422eb7876f973f191da98cc4c02aecbd0dbaaeb;hb=HEAD) from the wireshark repository
* A [dump](https://www.cloudshark.org/captures/36822505d7ab) on cloudshark

sgso commented:
@authmillenon Sure, I asked for recommendations regarding real hardware several times before but somehow always got ignored. So, what is the premium standard for running RIOT master as of today on real hardware? (if it matters: I have flashers lying around for Atmel, STM32 & LPC, comfortable with setting up AVR hardware, but can buy too). And I'd need some radio adapter for the PC as well.

authmillenon commented:
Oh sorry for ignoring you. Were did you post those questions? The best address is normally the devel list or the IRC. Anyway, I think the [SAMR21 xplained pro](http://www.atmel.com/tools/atsamr21-xpro.aspx) board is currently the best supported for a reasonable price.

authmillenon commented:
(we might even be able to provide you with 1 or 2 of these boards, just ask @emmanuelsearch or @haukepetersen).

authmillenon commented:
Rebased to current master and reordered a little bit.

Lotterleben commented on sys/include/net/ng_zep.h:
```diff
@@ -0,0 +1,234 @@
+ * @extends ng_netdev_t
+ */
+typedef struct {
+    ng_netdev_driver_t *driver;     /**< pointer to the devices interface */
```
nitpick: missing “ ' ” in devices

authmillenon replied:
Done

Lotterleben commented:
Okay, a word of warning: this PR will take me a while... (@sgso, if you still want to add your comments, I think it would be appreciated :>)

authmillenon commented:
nP I think I need a while anyway to find all the bugs :>

authmillenon commented:
Rebased to current master

authmillenon commented:
Rebased to current master

authmillenon commented:
Can someone explain to me, why a wrong UDP checksum is calculated, but when I step through the calculation with a debugger the checksum comes out right?

authmillenon commented:
Fixed as best as I could and circumvent #3201 making the UDP payload one big packet snip. I think ZEP itself is stable now. It shows similar behavior as the boards now with being able to ping 1000 packets, but letting the receiver crash/get stuck after a while. As hinted: in contrast to the boards the receiver is crashing after a while on a `ng_pkt_len()` with a wild running next pointer.

authmillenon commented:
Rebased to current master and added some fixes.

authmillenon commented:
Rebased again and added a new fix.

haukepetersen commented on sys/shell/commands/sc_zep.c:
```diff
@@ -0,0 +1,93 @@
+        switch (res) {
+            case -EOVERFLOW:
+                puts("error: too many threads running");
+
```
break missing.

cgundogan replied:
switch statements should break by default and only slide through if specified by a keyword .. this would be much less error-prone :dart:

haukepetersen commented on sys/shell/commands/sc_zep.c:
```diff
@@ -0,0 +1,93 @@
+
+            case -ENOTSUP:
+                printf("error: dst_addr (%s) invalid\n", argv[1]);
+
```
here and below: break missing?!

authmillenon commented:
"Addressed" comments by removing the `zep_init` command altogether and auto initialize ZEP with `ff02::1` as destination.

authmillenon commented:
I'll revert that one, because it causes an endless loop through the whole stack (ZEP sends a MC packet to IPv6, IPv6 dispatches it through all its interfaces including ZEP, ZEP sends a MC packet with the previous MC packet encapsulated to IPv6, IPv6 dispatches it through all …). This is a problem that might occur with all encapsulation interfaces. I don't have a solution for this yet. Proposals are welcome :-)

Lotterleben commented:
I fear the situation hasn't changed that I can't do this one justice... Would anyone who's currently testing ZEP be willing to take the PR over?

authmillenon commented:
Rebased to current master.

PeterKietzmann commented on tests/zep/main.c:
```diff
@@ -0,0 +1,76 @@
+ * @{
+ *
+ * @file
+ * @brief       Test application for Xbee S1 network device driver
```
Is this test needed? If yes, is this the correct description?

PeterKietzmann replied:
Now I see you introduced this test not long ago. 1. the "old" shell_init produces a warning. 2. how can I use this test? I mean, pinging and so on...

authmillenon replied:
This test is supposed to work like the other device tests (`tests/ng_at86rf2xx`, `tests/xbee`, etc.) Will fix descripiton and `shell_init`

authmillenon replied:
How to use: type in `zep_init` as before, but use txtsnd instead of ping ;-)

authmillenon replied:
To ping (which would make this app basically equivalent to `ng_networking`) you have to add the module `ng_icmpv6_echo`

PeterKietzmann replied:
Thx for the explanation. Once again I made a mistake during compilation yesterday. That was the reason for my confusion :-)

OlegHahm commented on sys/include/net/ng_zep.h:
```diff
@@ -0,0 +1,235 @@
+/**
+ * @brief   Maximum possible packet size in byte
+ */
+#define NG_ZEP_MAX_PKT_LENGTH   (116)
```
Why is this smaller than IEEE 802.15.4's max packet length?

authmillenon replied:
Where is it defined, what is it?

OlegHahm replied:
I thought this defines the frame length, not the payload length.

OlegHahm commented on sys/include/net/ng_zep.h:
```diff
@@ -0,0 +1,235 @@
+ */
+#define NG_ZEP_MIN_CHANNEL      (11U)
+#define NG_ZEP_MAX_CHANNEL      (26U)
+#define NG_ZEP_DEFAULT_CHANNEL  (17U)
```
Why do we need to define these values here again? What makes them different from IEEE 802.15.4 defines?

authmillenon replied:
Where are those defined? Aren't there also versions of the at86… device that excepts different channels.

OlegHahm replied:
Yeah, I realized that there are not yet centralized.

haukepetersen replied:
Is there actually an issue for this open? Otherwise we probably should open one...

PeterKietzmann replied:
#3291

OlegHahm commented on sys/include/net/ng_zep.h:
```diff
@@ -0,0 +1,235 @@
+ *
+ * TODO: Read some global network stack specific configuration value
+ */
+#define NG_ZEP_DEFAULT_PANID    (0x0023)
```
Same question here.

OlegHahm commented:
Without looking at the code in detail: I think for testing it would be much simpler to enable ZEP for now only for native and use the host's networking capabilities for the underlying UDP transport.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1079 @@
+#ifdef MODULE_NG_SIXLOWPAN
+    dev->proto = NG_NETTYPE_SIXLOWPAN;
+#else
+    dev->proto = NG_NETTYPE_UNDEF;
```
Wouldn't this be misuse of ZEP?

authmillenon replied:
Is using an IEEE 802.15.4 device without ZEP a missuse of that device?

OlegHahm replied:
Clarified offline.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1079 @@
+static int _send(ng_netdev_t *netdev, ng_pktsnip_t *pkt)
+{
+    ng_zep_t *dev = (ng_zep_t *)netdev;
+    ng_pktsnip_t *ptr, *new, *hdr;
```
I think `new` is an error prone name particular for C++.

Kijewski replied:
Does not matter in a .c file.

OlegHahm replied:
Ok, but the name is still not very speaking.

OlegHahm replied:
Why not `zep_pkt` or `encapsulated`?

authmillenon replied:
From our coding conventions:

>     LOCAL variable names should be short, and to the point.  If you have
>     some random integer loop counter, it should probably be called "i".
>     Calling it "loop_counter" is non-productive, if there is no chance of it
>     being mis-understood.  Similarly, "tmp" can be just about any type of
>     variable that is used to hold a temporary value.

I argue, that `new` is to the point, since it is the newly generated packet.

haukepetersen replied:
I am with @OlegHahm here: I find the use of the `new` keyword as variable name quite disturbing (although it is technically speaking ok to use it). One anyoing thing is, that is screws up the syntax highlighting in some frequently used editors (Eclipse, sublime text)...

authmillenon replied:
Those text editors are inferior to the great and all-powerful vim! 

jk will fix.

authmillenon replied:
Fixed

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1079 @@
+
+    new = hdr;
+
+    hdr = ng_ipv6_hdr_build(new, NULL, 0, (uint8_t *) & (dev->dst),
```
no space after address operator: `&(dev->dst)`

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1079 @@
+
+    new = hdr;
+
+    mhr_offset = _zep_hdr_fill(dev, zep, payload_len + hdr_len + 2);
```
What is the `2` standing for? Can we have a constant or macro?

authmillenon replied:
Length of the FCS field.

authmillenon commented:
> Without looking at the code in detail: I think for testing it would be much simpler to enable ZEP for now only for native and use the host's networking capabilities for the underlying UDP transport.

Huh? I'm somehow unable to make sense out of this sentence...

authmillenon commented:
Addressed comments that make sense to me.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1079 @@
+            case NG_NETAPI_MSG_TYPE_RCV:
+                DEBUG("zep: NG_NETAPI_MSG_TYPE_RCV\n");
+                ringbuffer_add(&_rx_buf, (char *)(&msg.content.ptr),
+                               sizeof(ng_pktsnip_t *));
```
nitpicky: but I think the indentation is off by one.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1079 @@
+    }
+
+    /* fill in destination PAN ID */
+    pos = 3;
```
3?

authmillenon replied:
See https://github.com/RIOT-OS/RIOT/pull/2627/files#diff-025cb696aa09ce44b700ad858efafe1fR872

OlegHahm replied:
related to #3291

OlegHahm commented on sys/net/network_layer/ng_ipv6/netif/ng_ipv6_netif.c:
```diff
@@ -527,7 +527,7 @@ void ng_ipv6_netif_init_by_dev(void)
             (if_type == NG_NETTYPE_SIXLOWPAN)) {
             DEBUG("Set 6LoWPAN flag\n");
-            ipv6_ifs->flags |= NG_IPV6_NETIF_FLAGS_SIXLOWPAN;
+            ipv6_ifs[i].flags |= NG_IPV6_NETIF_FLAGS_SIXLOWPAN;
```
How is this related to ZEP?

authmillenon replied:
Good point: see #3281

authmillenon commented:
Rebased to current master and adapted for simplifications in #2893.

authmillenon commented:
Added stability.

authmillenon commented:
Rebased again to include #3285

OlegHahm commented on sys/include/net/ng_zep.h:
```diff
@@ -0,0 +1,236 @@
+/**
+ * @defgroup    net_ng_zep  Zigbee Encapsulation Protocol
+ * @ingroup     net
+ * @brief       Transports IEEE 802.15.4 frames over UDP (can be parsed by
```
Maybe add a link to Wireshark documentation.

authmillenon replied:
Done

OlegHahm commented on sys/include/net/ng_zep.h:
```diff
@@ -0,0 +1,236 @@
+ * @brief   Default message queue size to use for the ZEP thread.
+ */
+#ifndef NG_ZEP_MSG_QUEUE_SIZE
+#define NG_ZEP_MSG_QUEUE_SIZE   (8U)
```
Are we sure that this is big enough? Since we're only using this feature on native currently, we don't have to be modest with memory here.

authmillenon replied:
What do you think is a big enough value?

OlegHahm replied:
I dunno, but with 32 or 64 we should be on the safe side, I would guess.

authmillenon replied:
Isn't this a little big, considering that this should emulate real IEEE 802.15.4 device drivers?

OlegHahm replied:
Let's get this working first and then think about realism (which is probably difficult for ZEP anyway) and optimization. Better be safe than sorry.

haukepetersen replied:
Actually, I would say if 8 or 64 does probably not make too much of a difference. If there is data coming in/going out faster then it is progressed, either sized message queue will fill up fast...

OlegHahm replied:
For fragmentation this might be relevant.

OlegHahm commented on sys/include/net/ng_zep.h:
```diff
@@ -0,0 +1,236 @@
+ * @brief   Option flags for the ZEP device
+ * @{
+ */
+#define NG_ZEP_FLAGS_AUTOACK            (0x0001)    /**< auto ACKS active */
```
Who's actually in charge of sending the ACK for ZEP?

authmillenon replied:
Some future implementation of this mechanism ;-)

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1072 @@
+
+    dev->driver = (ng_netdev_driver_t *)&_zep_driver;
+    dev->chan = NG_ZEP_DEFAULT_CHANNEL;
+    dev->pan = byteorder_btols(byteorder_htons(NG_ZEP_DEFAULT_PANID));
```
Why double converting?

authmillenon replied:
to get little-endian for sure (if we are not on a little-endian system).

OlegHahm replied:
Why does it has to be little-endian?

authmillenon replied:
Because it's IEEE 802.15.4 and I don't want to convert it for every packet.

OlegHahm replied:
But it is a constant value anyway and why do you convert it to host byte order first?

authmillenon replied:
I convert it to *network* byteorder (big-endian) first. I do this because there is no host-to-little-endian function set currently.

OlegHahm replied:
I still don't get this. You're basically swapping the byte-order (of a constant) twice.

authmillenon replied:
A constant that a user is able to change. Think like this:

* if we are on a big-endian system `byteorder_htons()` has no effect and `byteorder_btols()` swaps it to little-endian
* if we are on a little-endian system `byteorder_htons()` swaps it to big-endian and then `byteorder_btols()` back to little-endian.

Since it *is* a constant the compiler normally optimizes this out, so I thought I'm on the safer side here.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1072 @@
+    dev->proto = NG_NETTYPE_UNDEF;
+#endif
+
+    dev->seq = genrand_uint32();
```
Does ZEP require a randomly initialized sequence number?

authmillenon replied:
I thought it wouldn't hurt.

OlegHahm replied:
Well, if it pulls in another module that is otherwise not needed, maybe it does.

authmillenon replied:
For future usage this is definitely how it should be.

In the current configuration `random` is also included for `ng_ndp`, too. If we have another random-less network layer in the future we can discuss this further.

OlegHahm replied:
> For future usage this is definitely how it should be.

Why?

authmillenon replied:
Because sequence numbers shouldn't generally be some fixed value to begin with… Can we focus on some more pressing issues?

OlegHahm replied:
I cannot estimate how pressing this is, since I don't know the side effects.

authmillenon replied:
Me neither, but as stated: With IPv6 as network layer we can't get rid of `random` anyways...

OlegHahm replied:
But it would make testing/sniffing more reproducable if the checksum doesn't change.

authmillenon replied:
Changes too because of timestamp.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1072 @@
+        return -ENOBUFS;
+    }
+
+    zep = new->data;
```
Not a big issue, but wouldn't it be easier readable if you make this assignment just before you need it, i.e. at line 261?

authmillenon replied:
As you can see, I reassign `new` in l248 and l260 to save stack space, so moving this below l260 would result in an error.

OlegHahm replied:
Ah, yes, sorry.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1072 @@
+    ng_zep_hdr_t *zep;
+    size_t payload_len = ng_pkt_len(pkt->next), hdr_len, mhr_offset;
+    ng_netreg_entry_t *entry;
+    uint8_t mhr[NG_IEEE802154_MAX_HDR_LEN], *data;
```
What does `mhr` stand for?

authmillenon replied:
MAC header… I think. It's an abbreviation from the IEEE 802.15.4 spec.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1072 @@
+    return sizeof(ng_zep_v1_hdr_t);
+}
+
+static size_t _zep_hdr_fill_v2_data(ng_zep_t *dev, ng_zep_v2_data_hdr_t *hdr,
```
nitpicky: why does this function have a different naming scheme in comparison with the `_v1` one?

authmillenon replied:
because in v2 ACK frames and DATA frames have different ZEP headers.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1072 @@
+    uint32_t epoch_sec;
+
+    vtimer_get_localtime(&time);
+    epoch_sec = (uint32_t)mktime(&time);
```
Why not `vtimer_now()`?

authmillenon replied:
Because vtimer_now() does not return the current UNIX time, but the seconds since initialization (?)

OlegHahm replied:
Seconds since initialization is exactly what you get here.

OlegHahm replied:
Because of filling the tm struct by `vtimer_get_localtime()`.

authmillenon replied:
I hope for some future vtimer/wtimer implementation where I get the real local time ;-).

OlegHahm replied:
Why not calling `get_localtime()` directly instead? IIRC `vtimer_get_locatime()` was more a workaround anyway.

authmillenon replied:
Where is this defined?

authmillenon replied:
I guess you've meant `localtime()`... will do.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1072 @@
+
+void *_event_loop(void *args)
+{
+    msg_t msg, ack, msg_q[NG_ZEP_MSG_QUEUE_SIZE];
```
https://github.com/RIOT-OS/RIOT/issues/2851

authmillenon replied:
This issue is closed…

OlegHahm replied:
Did you read it?

authmillenon replied:
Yes, since all other event loops in the network stack are allocated this way, I thought we came to the conclusion to do it this way.

OlegHahm replied:
No.

haukepetersen replied:
Actually, I disagree with #2851! In this case, it makes IMHO much sense, to put the msg queue on the stack: just making a global static variable does not work here, as you could have multiple instances of ZEP devices. Same situation for `ng_nomac` and so on. As you don't necessarily know during compile time, how many instances there will be, allocating this on the stack is the most elegant solution!

OlegHahm replied:
What's the usecase of having multiple ZEP instances?

OlegHahm replied:
(And if you don't know how many instances of ZEP/nomac/... you have at compile time, how do you allocate memory for their stacks?)

authmillenon commented:
Addressed more comments.

authmillenon commented:
And even more.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1067 @@
+        return NULL;
+    }
+
+    pkt = ng_pktbuf_remove_snip(pkt, pkt);  /* remove FCS */
```
What the effect of given pkt as parameter twice? From the pktbuf API documentation I would assume that this removes the whole packet.

authmillenon replied:
From my point of view the documentation is clear on that point (shout loudly if you still disagree):

>      * @brief   Deletes a snip from a packet and the packet buffer.

So the snip `pkt` gets removed from the packet starting with snip `pkt` => it removes the first snip in the packet.

authmillenon replied:
(which is the frame-trailing FCS btw).

OlegHahm replied:
Well, it is at least a bit confusing that `pkt` is used as a "packet" and a "snip" at the same time.

authmillenon replied:
From ng_pkt.h:

>      * @brief   Type to represent snips (or parts,
>      *          snip == (packet header || packet payload)) of a network packet
>      * @details The idea behind the packet snips is that they either can represent
>      *          protocol-specific headers or payload. A packet can be comprised of
>      *          @f$ n @f$ pktsnip_t elements, where the first element represents the
>      *          header of the lowest available network layer and the @f$ (n - 1) @f$-st
>      *          element represents the payload of the highest available layer.
>      *          Use @ref sys_ut to operate on this.

I'm not in favor of explaining this everytime a function touches a packet...

OlegHahm replied:
Me neither, but we should extend the documentation at some point. The packet buffer is a complex, powerful tool and not always self-explanatory.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1067 @@
+
+    netif = _make_netif_hdr(mhr->data);
+
+    pkt = ng_pktbuf_remove_snip(pkt, mhr);
```
Why is this removed again?

authmillenon replied:
Since there are two versions of how to read your sentence:

1. *(Why is MHR removed again here)* It isn't, it is removed for the first time here.
2. *(I don't understand, why do you remove mhr)* Upper layers expect the general netif header, not the link-layer's specific header. The netif header is added in l738, the specific link-layer header (the MHR) is removed here.

OlegHahm replied:
I thought it was added in l726, but I somehow misread.

OlegHahm commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1067 @@
+    return pos;
+}
+
+static size_t _get_frame_hdr_len(uint8_t *mhr)
```
I cannot really review this function because I have no idea where the numbers come from.

authmillenon replied:
Me neither… As I said: I copied this from the then current version of Hauke's [`ng_at86rf2xx`](https://github.com/RIOT-OS/RIOT/blob/master/drivers/ng_at86rf2xx/ng_at86rf2xx_netdev.c#L105), since they generate the desired IEEE 802.15.4 header as far as I can tell.

OlegHahm commented:
I don't understand everything and I think there might still be issues, but I couldn't see anything serious for now. Hence, I give this an ACK.

authmillenon commented:
Squashed. Only merge conflict was regarding the change in 6eb6367.

authmillenon commented:
WTF … okay somehow the history is broken… Will fix

authmillenon commented:
Somehow the commit that is now d5d4b3f was deleted when I squashed in bulk. Now that I squashed and reordered step by step everything is fine. The statement to merge conflicts still upholds, its SHA is now 089bbdb

authmillenon commented:
PR needed some fixing. Can I have a re-ACK?

OlegHahm commented:
re-ACK, but don't merge yet. @haukepetersen wanted to review, too.

haukepetersen commented:
Japp, doing it as we speak

authmillenon commented:
Squashed and fixed the disconcerning naming of `new` to `new_pkt`.

haukepetersen commented on sys/net/application_layer/ng_zep/ng_zep.c:
```diff
@@ -0,0 +1,1067 @@
+    dev->dst.u64[0] = dst->u64[0];
+    dev->dst.u64[1] = dst->u64[1];
+    dev->dst_port = dst_port;
+    dev->version = 2;
```
is the version expected to change? Otherwise we could throw it out and make it constants?!

authmillenon replied:
Not in the current version. An older version of this implementation send version 1 packets and it is still possible with an easy update of `ng_netconf.h` (and setting this variable), which I neglected for now.

TL;DR changing the version is possible, but not implemented yet though.

haukepetersen replied:
I see

authmillenon commented:
Rebased to current master to simplify testing and updated for changes in #3159

haukepetersen commented:
Code wise I am fine with the PR for now. Though it does not work on my system: When rebasing on the newest master, initialize zep on two nodes and ping from one to the other, the receiver crashes (segfaults) when receiving neighbor solicitations over the ZEP interface. I put this in our [debugging list](https://docs.google.com/spreadsheets/d/1-sLRhkPiixjAAE4Yhx9Z9VWMuwz3LxGx7KIgkf9xMqI/edit#gid=0)... 

So if this PR is working for other people, I am ok with merging ZEP as is and fixing the findings in a next step. If it is working for no-one, it needs to be fixed first! :-)

authmillenon commented:
It works for me and @PeterKietzmann in the most current version (which is already rebased to the most current master) so what's the verdict? Should I squash?

haukepetersen commented:
ack

authmillenon commented:
Squashed

authmillenon commented:
\o/
